### PR TITLE
OS/Windows: Initial support for Windows

### PR DESCRIPTION
**Pull request by flavioislima**
Here is a list of things that I'm working right now:
- [x] Sign In
- [x] Load Library
- [x] Install Games
- [x] Launch Games
- [x] Import Games
- [x] Game Statuses (Progress is working perfectly)
- [x] Remove unneeded settings.
- [ ] Stop Installation (Have no idea on how to do it - Will do on another PR, maybe using AbortController introduced on Node14.7 and available on Electron 13).

Problems that need fixes:
- [x] A Better way to handle paths. On Windows, if the path contains `'`, legendary cannot do anything, so for now will need to use a replace function when iswindows.
- [x] Test a lot on all platforms to check what could go wrong.
- [ ] Selective download is not working so its not possible to download Fortnite or Cyberpunk yet. (Something related to conversion between cp1252 and utf8, will fix on another PR).

### CODE REVIEW

File: electron/main.ts [hunk 1]
```diff
@@ -1,36 +1,38 @@
...
+} from 'electron';
+import {
+  cpus,
+  platform
+} from 'os';
 import {
   exec,
   spawn
-} from 'child_process'
+} from 'child_process';
```
Comment (adityaruplaha) [inline electron/main.ts:20]: No semicolons please.

File: electron/main.ts [hunk 2]
```diff
@@ -52,10 +54,11 @@ import {
   loginUrl,
   sidInfoUrl,
   supportURL
-} from './constants'
-import { handleProtocol } from './protocol'
+} from './constants';
```
Comment (adityaruplaha) [inline electron/main.ts:57]: Ditto.

File: electron/main.ts [hunk 3]
```diff
@@ -52,10 +54,11 @@ import {
...
+import { handleProtocol } from './protocol';
```
Comment (adityaruplaha) [inline electron/main.ts:58]: Ditto.

File: electron/games.ts [hunk 4]
```diff
@@ -264,6 +267,17 @@ class LegendaryGame implements Game {
...
+      launcherArgs = ''
+
+    } = await this.getSettings()
+    const command = `${legendaryBin} launch ${this.appName} ${launcherArgs}`
+    console.log('\n Launch Command:', command)
+
+    return await execAsync(command)
+  }
+
   public async launch() {
```
Comment (adityaruplaha) [inline electron/games.ts:281]: I would suggest checking `isWindows` here instead of the caller side.

File: electron/legendary_utils/library.ts [hunk 5]
```diff
@@ -388,4 +411,4 @@ class Library {
   }
 }
 
-export { Library }
+export { Library };
```
Comment (adityaruplaha) [inline electron/legendary_utils/library.ts:414]: Semicolon.

File: electron/main.ts [hunk 6]
```diff
@@ -430,6 +435,10 @@ ipcMain.handle('refreshLibrary', async () => {
 ipcMain.handle('launch', (event, game) => {
   console.log('launching', game)
 
+  if (isWindows) {
```
Comment (adityaruplaha) [inline electron/main.ts:438]: Should not be necessary with what I said above.

File: electron/main.ts [hunk 7]
```diff
@@ -501,32 +510,50 @@ ipcMain.handle('updateGame', async (e, game) => {
 
 // TODO(adityaruplaha): Use UNIX sockets to refactor this.
 ipcMain.handle('requestGameProgress', async (event, appName) => {
```
Comment (adityaruplaha) [inline electron/main.ts:512]: Deliberately not reviewing this code as #385 will make it irrelevant.

File: electron/utils.ts [hunk 8]
```diff
@@ -1,24 +1,36 @@
...
+} from './constants';
 
 const execAsync = promisify(exec)
 const statAsync = promisify(stat)
 
 const { showErrorBox, showMessageBox } = dialog
 
 async function isOnline() {
-  return net.isOnline
+  return net.isOnline()
```
Comment (adityaruplaha) [inline electron/utils.ts:27]: Shit.

File: src/screens/Game/GamePage/index.tsx [hunk 9]
```diff
@@ -14,17 +22,25 @@ import {
...
-import Settings from '@material-ui/icons/Settings'
-import UpdateComponent from 'src/components/UI/UpdateComponent'
+} from 'src/helpers';
+import { useParams } from 'react-router-dom';
+import { useTranslation } from 'react-i18next';
+import ContextProvider from 'src/state/ContextProvider';
+import Header from 'src/components/UI/Header';
+import InfoBox from 'src/components/UI/InfoBox';
+import UpdateComponent from 'src/components/UI/UpdateComponent';
+/* eslint-disable complexity */
```
Comment (adityaruplaha) [inline src/screens/Game/GamePage/index.tsx:32]: This should be unnecessary?

File: src/screens/Library/components/GameCard/index.tsx [hunk 10]
```diff
@@ -1,7 +1,11 @@
 import './index.css'
 
 /* eslint-disable complexity */
```
Comment (adityaruplaha) [inline src/screens/Library/components/GameCard/index.tsx:3]: This should be dropped?

File: src/screens/Settings/components/OtherSettings/index.tsx [hunk 11]
```diff
@@ -1,9 +1,12 @@
-import React, { ChangeEvent } from 'react'
+import React, {
+  ChangeEvent,
+  useContext
+} from 'react';
```
Comment (adityaruplaha) [inline src/screens/Settings/components/OtherSettings/index.tsx:4]: Semicolons.

File: src/state/GlobalState.tsx [hunk 12]
```diff
@@ -61,6 +61,8 @@ export class GlobalState extends PureComponent<Props> {
     const { user, library } = await getLegendaryConfig()
     const updates = await renderer.invoke('checkGameUpdates')
 
+    console.log(process.platform)
```
Comment (adityaruplaha) [inline src/state/GlobalState.tsx:64]: Debug stuff you forgot to remove? :P

File: src/state/ContextProvider.tsx [hunk 13]
```diff
@@ -1,5 +1,7 @@
 import React from 'react'
 
+const { remote } = window.require('electron')
+const { process } = remote
```
Comment (adityaruplaha) [inline src/state/ContextProvider.tsx:4]: Can be combined into a single line I guess.

File: package.json [hunk 14]
```diff
@@ -91,13 +97,14 @@
     "start": "nf start -p 3000 && electron-start",
     "electron-start": "ELECTRON_IS_DEV=1 node public/start-react",
     "electron": "yarn build-electron && electron .",
-    "react-start": "BROWSER=none react-scripts start",
+    "react-start": "react-scripts start",
```
Comment (adityaruplaha) [inline package.json:100]: Does this change do anything?

File: electron/utils.ts [hunk 15]
```diff
@@ -114,4 +126,4 @@ export {
   openUrlOrFile,
   showAboutWindow,
   statAsync
-}
+};
```
Comment (adityaruplaha) [inline electron/utils.ts:129]: Semicolon.

File: electron/main.ts [hunk 16]
```diff
@@ -39,10 +41,10 @@ import {
   isOnline,
   openUrlOrFile,
   showAboutWindow
-} from './utils'
-import { dialog } from 'electron'
+} from './utils';
```
Comment (adityaruplaha) [inline electron/main.ts:44]: Ditto.

File: electron/main.ts [hunk 17]
```diff
@@ -1,36 +1,38 @@
...
+import Backend from 'i18next-fs-backend';
+import i18next from 'i18next';
+import isDev from 'electron-is-dev';
+
+import { DXVK } from './dxvk';
+import { GameConfig } from './game_config';
+import { GlobalConfig } from './config';
+import { LegendaryGame } from './games';
+import { Library } from './legendary_utils/library';
 import { User } from './legendary_utils/user';
```
Comment (adityaruplaha) [inline electron/main.ts:36]: Ditto, lots of semicolons above.

File: electron/games.ts [hunk 18]
```diff
@@ -199,12 +200,14 @@ class LegendaryGame implements Game {
   public async install(path : string) {
     const { maxWorkers } = (await GlobalConfig.get().getSettings())
     const workers = maxWorkers === 0 ? '' : `--max-workers ${maxWorkers}`
-
+    // Need to fix convertion from utf8 to win1252 or vice-versa
```
Comment (adityaruplaha) [inline electron/games.ts:203]: I would recommend marking this as an official `//TODO(flavioislima):`

File: electron/constants.ts [hunk 19]
```diff
@@ -52,12 +74,12 @@ export {
...
   legendaryBin,
   legendaryConfigPath,
   libraryPath,
   loginUrl,
-  shell,
   sidInfoUrl,
   supportURL,
   userInfo
-}
+};
```
Comment (adityaruplaha) [inline electron/constants.ts:85]: Semicolon.

**Review by adityaruplaha — CHANGES_REQUESTED** [review]
Lots of minor changes are necessary imo, mostly for style conformance.

File: electron/main.ts (hunk 1 above)
Comment (flavioislima) [reply electron/main.ts:20]: On Windows my settings are different than on Linux so its putting this when auto-import. Will fix.

File: .github/workflows/test.yml [hunk 20]
```diff
@@ -13,8 +13,4 @@ jobs:
       uses: actions/checkout@v2
     - name: Install modules.
       run: yarn
-    - name: Test and collect coverage.
-      uses: artiomtr/jest-coverage-report-action@v1.3
-      with:
-        github_token: ${{ secrets.GITHUB_TOKEN }}
-        test_script: yarn test --coverage
+    - run: yarn test:ci
```
Comment (adityaruplaha) [inline .github/workflows/test.yml:16]: This looks kinda out of scope for this PR.

File: .husky/post-merge [hunk 21]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-yarn
+npm install
```
Comment (adityaruplaha) [inline .husky/post-merge:4]: This should be its own PR tbh, if we decide to replace yarn with npm. Mixing is no fun.

File: .husky/pre-commit [hunk 22]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-yarn lint-fix
+npm run lint-fix
```
Comment (adityaruplaha) [inline .husky/pre-commit:4]: Ditto.

File: .husky/pre-push [hunk 23]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-yarn lint
+npm run lint
```
Comment (adityaruplaha) [inline .husky/pre-push:4]: Ditto.

File: package.json [hunk 24]
```diff
@@ -87,14 +93,17 @@
   "scripts": {
     "start": "nf start -p 3000 && electron-start",
     "electron-start": "ELECTRON_IS_DEV=1 node public/start-react",
-    "electron": "yarn build-electron && electron .",
-    "react-start": "BROWSER=none react-scripts start",
+    "electron": "npm run build-electron && electron .",
```
Comment (adityaruplaha) [inline package.json:96]: Same as I said above.

File: src/screens/Game/GamePage/index.tsx [hunk 25]
```diff
@@ -1,8 +1,16 @@
-import './index.css'
+import './index.css';
```
Comment (adityaruplaha) [inline src/screens/Game/GamePage/index.tsx:1]: You missed a semicolon. :P

File: src/screens/Game/GamePage/index.tsx [hunk 26]
```diff
@@ -509,12 +525,16 @@ export default function GamePage(): JSX.Element | null {
       }
 
       if (installPath === 'import' && gameInfo.is_game) {
-        const { filePaths } = await showOpenDialog({
+        const { filePaths, canceled } = await showOpenDialog({
```
Comment (adityaruplaha) [inline src/screens/Game/GamePage/index.tsx:528]:
```suggestion
        const { filePaths, cancelled } = await showOpenDialog({
```

File: src/screens/Game/GamePage/index.tsx [hunk 27]
```diff
@@ -524,12 +544,16 @@ export default function GamePage(): JSX.Element | null {
       }
 
       if (installPath === 'another' || !gameInfo.is_game) {
-        const { filePaths } = await showOpenDialog({
+        const { filePaths, canceled } = await showOpenDialog({
```
Comment (adityaruplaha) [inline src/screens/Game/GamePage/index.tsx:547]:
```suggestion
        const { filePaths, cancelled } = await showOpenDialog({
```

File: src/screens/Game/GamePage/index.tsx [hunk 28]
```diff
@@ -524,12 +544,16 @@ export default function GamePage(): JSX.Element | null {
...
           buttonLabel: t('box.choose'),
           properties: ['openDirectory'],
           title: t('box.installpath')
         })
 
+        if (canceled) {
```
Comment (adityaruplaha) [inline src/screens/Game/GamePage/index.tsx:553]:
```suggestion
        if (cancelled) {
```

File: src/screens/Settings/index.tsx [hunk 29]
```diff
@@ -100,7 +118,7 @@ function Settings() {
     cloudSaveEnabled: false,
     saveFolder: ''
   })
-  const [autoSyncSaves, setAutoSyncSaves] = useState(false)
+  const [autoSyncSaves, setAutoSyncSaves] = useState(true)
```
Comment (adityaruplaha) [inline src/screens/Settings/index.tsx:121]: Is this safe to enable by default now?

**Review by adityaruplaha — COMMENTED** [review]
More comments.

File: .github/workflows/test.yml (hunk 20 above)
Comment (flavioislima) [reply .github/workflows/test.yml:16]: Yeja but without this change the tests weren't passing.

File: .husky/post-merge (hunk 21 above)
Comment (flavioislima) [reply .husky/post-merge:4]: But we did decide. On ci and scripts should be npm.

File: .husky/pre-commit (hunk 22 above)
Comment (flavioislima) [reply .husky/pre-commit:4]: Ditto

File: package.json (hunk 24 above)
Comment (flavioislima) [reply package.json:96]: Same for me. Scripts let's use npm 😝

File: .github/workflows/test.yml (hunk 20 above)
Comment (adityaruplaha) [reply .github/workflows/test.yml:16]: Well should be fixed after #400.

File: src/screens/Game/GamePage/index.tsx (hunk 26 above)
Comment (flavioislima) [reply src/screens/Game/GamePage/index.tsx:528]: That's weird because this came from the electron api. So it might be wrong on their side. Changing it to cancelled doesn't work.

File: .husky/post-merge (hunk 21 above)
Comment (adityaruplaha) [reply .husky/post-merge:4]: Alright, but this looks kinda out of scope for this PR. I'll backport to a new one.

File: src/screens/Settings/index.tsx (hunk 29 above)
Comment (flavioislima) [reply src/screens/Settings/index.tsx:121]: Yes, safe because we have another check on the game side. If the game has no cloud save then it won't check that.

Comment (flavioislima) [reply src/screens/Settings/index.tsx:121]: Actually I should revert this. Because we don't have the folder until the user goes to sync settings.

File: src/screens/Game/GamePage/index.tsx (hunk 26 above)
Comment (flavioislima) [reply src/screens/Game/GamePage/index.tsx:528]: Yeah, check https://www.electronjs.org/docs/api/dialog#dialogshowopendialogsyncbrowserwindow-options

File: .github/workflows/test.yml (hunk 20 above)
Comment (flavioislima) [reply .github/workflows/test.yml:16]: Yeah but it weren't and I was trying to find the issue 😅
Anyway, I don't think will conflict.

File: src/screens/Game/GamePage/index.tsx (hunk 25 above)
Comment (flavioislima) [reply src/screens/Game/GamePage/index.tsx:1]: Damn semicolons!

File: src/screens/Game/GamePage/index.tsx (hunk 26 above)
Comment (adityaruplaha) [reply src/screens/Game/GamePage/index.tsx:528]: Weird, they made a typo I guess.

File: src/screens/Settings/index.tsx (hunk 29 above)
Comment (adityaruplaha) [reply src/screens/Settings/index.tsx:121]: Yeah I thought so too.

File: .github/workflows/test.yml (hunk 20 above)
Comment (adityaruplaha) [reply .github/workflows/test.yml:16]: Anyways, its safe to revert, #400 will fix this.

File: src/screens/Game/GamePage/index.tsx (hunk 26 above)
Comment (flavioislima) [reply src/screens/Game/GamePage/index.tsx:528]: or no:
`canceled or cancelled is the past tense of the verb to cancel. Both spellings are correct; Americans favor canceled (one L), while cancelled (two Ls) is preferred in British English and other dialects. `

File: electron/games.ts (hunk 4 above)
Comment (flavioislima) [reply electron/games.ts:282]: Ok

File: package.json (hunk 14 above)
Comment (flavioislima) [reply package.json:97]: I shouln't commit this. On windows it gives an error, I think we will need a separate react-start for windows.

File: src/state/ContextProvider.tsx (hunk 13 above)
Comment (flavioislima) [reply src/state/ContextProvider.tsx:4]: ACtually, remote was deprecated since Electron 12, so there is a lot of things to do to migrate to the new @electron/remote.
I'm just not sure if I want to do it in this PR

Comment (flavioislima) [reply src/state/ContextProvider.tsx:4]: https://github.com/electron/remote

File: src/screens/Game/GamePage/index.tsx (hunk 9 above)
Comment (flavioislima) [reply src/screens/Game/GamePage/index.tsx:32]: yes

File: src/screens/Game/GamePage/index.tsx (hunk 26 above)
Comment (adityaruplaha) [reply src/screens/Game/GamePage/index.tsx:528]: Damn didn't know about this one, I generally always right US English while coding, and English (India) normally. Noted. 👍🏾

File: electron/games.ts [hunk 30]
```diff
@@ -58,6 +60,9 @@ class LegendaryGame implements Game {
    * Alias for `Library.listUpdateableGames`
    */
   public static async checkGameUpdates() {
+    if (!User.isLoggedIn()){
```
Comment (adityaruplaha) [inline electron/games.ts:63]: This is no longer a true alias in that case, would recommend modifying the original function.

File: package.json [hunk 31]
```diff
@@ -123,7 +133,7 @@
     "@typescript-eslint/eslint-plugin": "^4.14.0",
     "@typescript-eslint/parser": "^4.14.0",
     "electron": "^13.0.1",
-    "electron-builder": "^22.11.3",
+    "electron-builder": "^22.10.5",
```
Comment (adityaruplaha) [inline package.json:136]: Why was this changed?

File: src/screens/Settings/index.tsx (hunk 29 above)
Comment (adityaruplaha) [reply src/screens/Settings/index.tsx:121]: This isn't fixed yet?

**Review by adityaruplaha — APPROVED** [review]
Looks good mostly, got a couple of small things left.

File: package.json (hunk 31 above)
Comment (flavioislima) [reply package.json:136]: Because this is the stable version. The other one is beta and is not creating releases automatically.

**Comment by flavioislima** [comment]
> Looks good mostly, got a couple of small things left.

OK, I'll fix them and merge it.
The autosync change is in two places and I forgot that one.
I also just found a new bug where the filter for updates is showing all games, I think it was after merging in master.

File: src/screens/Settings/index.tsx (hunk 29 above)
Comment (flavioislima) [reply src/screens/Settings/index.tsx]: Fixed.
